### PR TITLE
Add NominatedNodeName field to PodStatus

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**:
Today, Scheduler uses an annotation called "nominated-node-name" to mark a preemptor Pod. This annotation helps scheduler know about the Pods that are destined to run on the nodes so that the resources made available by preemption is not allocated to a different Pod. In a recent discussion with @bgrant0607, we learned that we should change the annotation to a field as this field can be used by multiple schedulers and other components that may make scheduling-related decisions (descheduler, auto-scaler, kube-arbitrator, ...). 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

ref #57471

**Special notes for your reviewer**:

**Release note**:

```release-note
Add "nominatedNodeName" field to PodStatus. This field is set when a pod preempts other pods on the node.
```
/sig scheduling

### CODE REVIEW

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -3051,6 +3051,13 @@ type PodStatus struct {
 	// More info: https://git.k8s.io/community/contributors/design-proposals/node/resource-qos.md
 	// +optional
 	QOSClass PodQOSClass `json:"qosClass,omitempty" protobuf:"bytes,9,rep,name=qosClass"`
+	// NominatedNodeName is set when this pod preempts other pods on the node, but it cannot be
```
Comment (bgrant0607) [inline staging/src/k8s.io/api/core/v1/types.go]: These comments are translated to user-facing documentation. Either don't mention the field name or use the json field name (nominatedNodeName).

Comment (bgrant0607) [reply staging/src/k8s.io/api/core/v1/types.go]: Details to specify somewhere:
1. Will this field always be set before nodeName is set?
2. Will this field be unset when nodeName is set?
3. Will this field be changed if nodeName is set to something different?

**Comment by bgrant0607** [comment]
Note: New fields shouldn't necessarily be put at the bottom. Please place the field near logically related fields, in order of importance / relevance.

Will the annotation still be understood/respected in the next release?

**Comment by bgrant0607** [comment]
Also, there should be a release note with the introduction of any new API field.

**Comment by bgrant0607** [comment]
Release-note guidance: https://github.com/kubernetes/community/issues/484

File: staging/src/k8s.io/api/core/v1/types.go (hunk 1 above)
Comment (bsalamat) [reply staging/src/k8s.io/api/core/v1/types.go]: Fixed the comment and moved the field up. Also added one more line to the comment to mention that NominatedNodeName and NodeName are not necessarily the same. More details about how this field is set is provided in [preemption design doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/pod-preemption.md#starvation-problem).

**Comment by bgrant0607** [comment]
Please write the release note in a form that would match what a user would see in the API.

File: staging/src/k8s.io/api/core/v1/generated.proto [hunk 2]
```diff
@@ -3170,6 +3170,16 @@ message PodStatus {
   // +optional
   optional string reason = 4;
 
+  // nominatedNodeName is set only when this pod preempts other pods on the node, but it cannot be
+  // scheduled right away as preemption victims receive their graceful termination periods.
+  // This field does not guarantee that the pod will be scheduled on this node. Scheduler may decide
+  // to place the pod elsewhere if other nodes become available sooner. Scheduler may also decide to
+  // give the resources on this node to a higher priority pod that is created after preemption.
+  // As a result, this field may be different than PodSpec.nodeName when the pod is
```
Comment (bgrant0607) [inline staging/src/k8s.io/api/core/v1/generated.proto]: Why is it desirable to allow them to be different? Why not either set them to match or clear the nominatedNodeName once nodeName is set?

Comment (bsalamat) [reply staging/src/k8s.io/api/core/v1/generated.proto]: There are a couple of reasons:
1. It is useful to understand/debug behavior of the cluster. When a pod causes preemption, but it is scheduled elsewhere, checking nominatedNodeName reveals that this is the pod that caused  preemption on the node. Users who don't have access to scheduler logs can still see this field.
2. Performance: setting the nominatedNodeName to match the nodeName or clearing it requires another API call.

**Comment by bsalamat** [comment]
> Please write the release note in a form that would match what a user would see in the API.

Done.
Thanks for you review, @bgrant0607!

**Pull request merged**